README.md
![header.jpg](header.jpg)

## Important

This is a fork of original widget made specifically for my needs. What's changed:

- Removed useless dots at the top of DayItem
- Added a dot at the bottom of DayItem, which could indicate presence of any event on that date. It is controlled using dayWithEventsPredicate and dotsColor
- Added scale argument to make widget smaller or larger in certain scenarios

**Flutter widget form select a date in horizontal timeline with customizable styles.**

## Getting Started

You can use this package when you need to add a date picker that takes up little screen space, so we can always have it visible, and that facilitates use with one hand.

![example.gif](example.gif)

## Properties

| Property                 | Type                   | Description                                                 |
| :----------------------- | :--------------------- | :---------------------------------------------------------- |
| initialDate              | DateTime               | Initial date selected                                       |
| firstDate                | DateTime               | First date available in calendar                            |
| lastDate                 | DateTime               | Last date available in calendar                             |
| selectableDayPredicate   | SelectableDayPredicate | Signature for predicating dates for enabled date selections |
| dayWithEventsPredicate   | DayWithEventsPredicate | Controls if the bottom dot is present                       |
| onDateSelected           | OnDateSelected         | Callback to notify that a date has been selected            |
| leftMargin               | double                 | Left margin to month and day list                           |
| monthColor               | Color                  | Color for month list elements                               |
| dayColor                 | Color                  | Color for day list elements                                 |
| activeDayColor           | Color                  | Color for selected day text                                 |
| activeBackgroundDayColor | Color                  | Color for selected day background                           |
| dotColor                 | Color                  | Color for top dots in select day                            |
| locale                   | String                 | Locale string to get formatted date                         |
| showYears                | bool                   | Indicates if it show year selector                          |

## Use example

You can review the example folder for a complete example of using the widget.

```
CalendarTimeline(
  initialDate: DateTime(2020, 4, 20),
  firstDate: DateTime(2019, 1, 15),
  lastDate: DateTime(2020, 11, 20),
  onDateSelected: (date) => print(date),
  leftMargin: 20,
  monthColor: Colors.blueGrey,
  dayColor: Colors.teal[200],
  activeDayColor: Colors.white,
  activeBackgroundDayColor: Colors.redAccent[100],
  dotsColor: Color(0xFF333A47),
  selectableDayPredicate: (date) => date.day != 23,
  locale: 'en_ISO',
)
```


CHANGELOG.md
## 2.1.0 - 12/11/2021

* Added default colors
* Added scale decorator for month & year items
* Fix bug with event predicate & day item height

## 1.0.0-null-safety.0 - 19/03/2021

* Migrate to null safety

## 0.7.1 - 09/03/2021

* Update packages

## 0.7.0 - 17/12/2020

* Added option to see years individually. Thanks to https://github.com/Flucadetena.

## 0.6.3 - 10/08/2020

* Fix error when generating the abbreviation of the day of the week in different locales
* Show the selected day when regenerating the days after pressing a month
* Upgrade dependencies

## 0.6.2 - 26/06/2020

* Check local localizations for intl default config
* Allow the last item in the listings to be left aligned

## 0.6.1 - 17/06/2020

* Update widget if state changes

## 0.6.0 - 03/06/2020

* Added locale parameter

## 0.5.0 - 05/05/2020

* Add dayNameColor parameter

## 0.4.0 - 29/04/2020

* Modify design year item

## 0.3.1 - 23/04/2020

* Fix error if initialDate and firstDate are the same day

## 0.3.0 - 22/04/2020

* Add dotColor parameter
* Show year with 4 digits and not show it like first element in month list
* Fix design errors

## 0.2.0 - 22/04/2020

* Show year in first element of month list
* Add parameters description to readme
* Modify example design

## 0.1.0 - 21/04/2020

* Add color customization
* Add selectableDayPredicate functionality
* Complete example app
* Fix errors
* Refactor

## 0.0.1 - 16/04/2020

* Initial version.


lib/src/day_item.dart
import 'package:flutter/material.dart';

/// Creates a Widget representing the day.
class DayItem extends StatelessWidget {
  final int dayNumber;
  final String shortName;
  final bool isDimmed;
  final bool hasEvents;
  final bool available;
  final bool isSelected;
  final void Function()? onTap;
  final Color? dayColor;
  final Color? dotsColor;
  final Color? dayNameColor;
  final Color? activeDayColor;
  final Color? activeDayBackgroundColor;
  final double scale;
  final double height;
  final double width;
  final double? dayBorderRadius;

  DayItem({
    Key? key,
    required this.dayNumber,
    required this.shortName,
    required this.onTap,
    this.isDimmed = false,
    this.available = true,
    this.hasEvents = false,
    this.isSelected = false,
    this.scale = 1.0,
    this.dayColor,
    this.dotsColor,
    this.dayNameColor,
    this.activeDayColor,
    this.activeDayBackgroundColor,
    this.dayBorderRadius,
  })  : height = 70.0 * scale,
        width = 60.0 * scale,
        super(key: key);

  ///? I united both widgets to increase the touch target of non selected days by using a transparent box decorator.
  ///? Now if the user click close to the number but not straight on top it will still select the date. (ONLY INFORMATION - ERASE)
  _buildDay(BuildContext context) {
    final textStyle = TextStyle(
      color: available && !isDimmed
          ? dayColor ?? Theme.of(context).accentColor
          : dayColor?.withOpacity(0.5) ??
              Theme.of(context).accentColor.withOpacity(0.5),
      height: 1,
      fontSize: 32 * scale,
      fontWeight: FontWeight.normal,
    );

    final selectedStyle = TextStyle(
      color: activeDayColor ?? Colors.white,
      height: 1,
      fontSize: 32 * scale,
      fontWeight: FontWeight.bold,
    );

    return GestureDetector(
      onTap: available ? onTap : null,
      child: Container(
        width: width,
        height: height,
        alignment: Alignment.center,
        decoration: isSelected
            ? BoxDecoration(
                color:
                    activeDayBackgroundColor ?? Theme.of(context).accentColor,
                borderRadius: BorderRadius.circular(dayBorderRadius ?? 12.0),
              )
            : BoxDecoration(color: Colors.transparent),
        child: Column(
          children: <Widget>[
            SizedBox(height: 13.0 * scale),
            Text(
              dayNumber.toString(),
              style: isSelected ? selectedStyle : textStyle,
            ),
            Text(
              shortName,
              style: TextStyle(
                fontSize: 14 * scale,
                height: 1,
                fontWeight: FontWeight.w500,
                color: isDimmed || !available
                    ? (dayNameColor == null
                        ? Colors.white.withOpacity(0.5)
                        : dayNameColor!.withOpacity(0.5))
                    : dayNameColor ?? Colors.white,
              ),
            ),
            SizedBox(height: 4.0 * scale),
            if (hasEvents) _buildDot(),
            SizedBox(height: 6.0 * scale),
          ],
        ),
      ),
    );
  }

  Widget _buildDot() {
    return Container(
      width: 5 * scale,
      height: 5 * scale,
      decoration: new BoxDecoration(
        color: this.dotsColor ?? this.activeDayColor ?? Colors.white,
        shape: BoxShape.circle,
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return _buildDay(context);
  }
}


lib/src/year_item.dart
import 'package:flutter/material.dart';

/// Creates a Widget to represent the years. By default it will show the smaller version
/// in the months row. If [small] is set to false it will show the bigger version for the
/// years row. In the smaller version the [onTap] property is not available
class YearItem extends StatelessWidget {
  final String name;
  final void Function()? onTap;
  final bool isSelected;
  final Color? color;
  final bool small;
  final double scale;

  YearItem({
    required this.name,
    required this.onTap,
    this.isSelected = false,
    this.small = true,
    this.scale = 1.0,
    this.color,
  });

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onTap: small ? null : onTap,
      child: Container(
        padding: const EdgeInsets.only(
          left: 14.0,
          right: 14.0,
          top: 3.0,
          bottom: 2.0,
        ),
        decoration: isSelected || small
            ? BoxDecoration(
                border: Border.all(color: color ?? Colors.black87, width: 1),
                borderRadius: BorderRadius.circular(4),
              )
            : null,
        child: Text(
          name.toUpperCase(),
          style: TextStyle(
            fontSize: small ? (14 * scale) : (20 * scale),
            color: color ?? Colors.black87,
            fontWeight: FontWeight.bold,
            height: 1,
          ),
        ),
      ),
    );
  }
}


lib/src/calendar_timeline.dart
import 'package:flutter/material.dart';
import 'package:flutter/scheduler.dart';
import 'package:intl/date_symbol_data_local.dart';
import 'package:intl/intl.dart';
import 'package:scrollable_positioned_list/scrollable_positioned_list.dart';

import 'day_item.dart';
import 'month_item.dart';
import 'util/string_extension.dart';
import 'year_item.dart';

typedef OnDateSelected = void Function(DateTime?);

/// Creates a minimal, small profile calendar to select specific dates.
/// [initialDate] must not be [null], the same or after [firstDate] and
/// the same or before [lastDate]. [firstDate] must not be [null].
/// [lastDate] must not be null and the same or after [firstDate]
class CalendarTimeline extends StatefulWidget {
  final DateTime initialDate;
  final DateTime firstDate;
  final DateTime lastDate;
  final SelectableDayPredicate? selectableDayPredicate;
  final SelectableDayPredicate? dayWithEventsPredicate;
  final SelectableDayPredicate? dayActivePredicate;

  final OnDateSelected onDateSelected;
  final double leftMargin;
  final double offset;
  final Color? dayColor;
  final Color? activeDayColor;
  final Color? activeBackgroundDayColor;
  final Color? monthColor;
  final Color? dotsColor;
  final Color? dayNameColor;
  final String? locale;

  /// If true, it will show a separate row for the years.
  /// It defaults to false
  final bool showYears;

  final bool showMonths;

  final double scale;

  CalendarTimeline({
    Key? key,
    required this.initialDate,
    required this.firstDate,
    required this.lastDate,
    required this.onDateSelected,
    this.selectableDayPredicate,
    this.dayWithEventsPredicate,
    this.dayActivePredicate,
    this.leftMargin = 10,
    this.offset = 20,
    this.dayColor,
    this.activeDayColor,
    this.activeBackgroundDayColor,
    this.monthColor,
    this.dotsColor,
    this.dayNameColor,
    this.locale,
    this.showMonths = true,
    this.showYears = false,
    this.scale = 1.0,
  })  : assert(
          initialDate.difference(firstDate).inDays >= 0,
          'initialDate must be on or after firstDate',
        ),
        assert(
          !initialDate.isAfter(lastDate),
          'initialDate must be on or before lastDate',
        ),
        assert(
          !firstDate.isAfter(lastDate),
          'lastDate must be on or after firstDate',
        ),
        assert(
          selectableDayPredicate == null || selectableDayPredicate(initialDate),
          'Provided initialDate must satisfy provided selectableDayPredicate',
        ),
        assert(
          locale == null || dateTimeSymbolMap().containsKey(locale),
          'Provided locale value doesn\'t exist',
        ),
        super(key: key);

  @override
  _CalendarTimelineState createState() => _CalendarTimelineState();
}

class _CalendarTimelineState extends State<CalendarTimeline> {
  final ItemScrollController _controllerYear = ItemScrollController();
  final ItemScrollController _controllerMonth = ItemScrollController();
  final ItemScrollController _controllerDay = ItemScrollController();

  int? _yearSelectedIndex;
  int? _monthSelectedIndex;
  int? _daySelectedIndex;
  late double _scrollAlignment;

  late Color _dotsColor;
  late Color _dayColor;
  late Color _monthColor;
  late Color _activeDayColor;
  late Color _activeBackgroundDayColor;

  List<DateTime> _years = [];
  List<DateTime> _months = [];
  List<DateTime> _days = [];
  DateTime? _selectedDate;

  String get _locale =>
      widget.locale ?? Localizations.localeOf(context).languageCode;

  /// Populates the calendar and animates to the [widget.initialDate]
  @override
  void initState() {
    super.initState();
    _initCalendar();
    _scrollAlignment = widget.offset / 440;
    _dotsColor = widget.dotsColor ?? Colors.white;
    _dayColor = widget.dayColor ?? Colors.grey;
    _activeDayColor = widget.activeDayColor ?? Colors.white;
    _monthColor = widget.monthColor ?? Colors.white70;
    _activeBackgroundDayColor = widget.activeBackgroundDayColor ?? Colors.blue;
    SchedulerBinding.instance!.addPostFrameCallback((_) {
      initializeDateFormatting(_locale);
    });
  }

  @override
  void didChangeDependencies() {
    _dotsColor = widget.dotsColor ?? Theme.of(context).colorScheme.secondary;
    _dayColor = widget.dayColor ?? Theme.of(context).scaffoldBackgroundColor;
    _activeDayColor =
        widget.activeDayColor ?? Theme.of(context).scaffoldBackgroundColor;
    _monthColor =
        widget.monthColor ?? Theme.of(context).scaffoldBackgroundColor;
    _activeBackgroundDayColor = widget.activeBackgroundDayColor ??
        Theme.of(context).scaffoldBackgroundColor.withOpacity(0.15);
    super.didChangeDependencies();
  }

  /// Refreshes the calendar when a day, month or year is selected
  @override
  void didUpdateWidget(CalendarTimeline oldWidget) {
    super.didUpdateWidget(oldWidget);
    _initCalendar();
    if (widget.showYears) _moveToYearIndex(_yearSelectedIndex ?? 0);
    _moveToMonthIndex(_monthSelectedIndex ?? 0);
    _moveToDayIndex(_daySelectedIndex ?? 0);
  }

  @override
  Widget build(BuildContext context) {
    return Column(
      mainAxisAlignment: MainAxisAlignment.center,
      mainAxisSize: MainAxisSize.min,
      children: <Widget>[
        if (widget.showYears) _buildYearList(),
        if (widget.showMonths) _buildMonthList(),
        _buildDayList(),
      ],
    );
  }

  /// Creates the row with the day of the [selectedDate.month]. If the
  /// [selectedDate.year] && [selectedDate.month] is the [widget.firstDate] or [widget.lastDate]
  /// the days show will be the available
  Widget _buildDayList() {
    return Container(
      height: 75 * widget.scale,
      child: ScrollablePositionedList.separated(
        itemScrollController: _controllerDay,
        initialScrollIndex: _daySelectedIndex ?? 0,
        initialAlignment: _scrollAlignment,
        scrollDirection: Axis.horizontal,
        itemCount: _days.length,
        padding: EdgeInsets.only(
          left: widget.offset,
          right: widget.offset,
        ),
        separatorBuilder: (context, index) =>
            Divider(indent: widget.offset / 5),
        itemBuilder: (BuildContext context, int index) {
          final currentDay = _days[index];
          final shortName =
              DateFormat.E(_locale).format(currentDay).capitalize();

          return Container(
            child: DayItem(
              isSelected: widget.dayActivePredicate == null
                  ? _daySelectedIndex == index
                  : widget.dayActivePredicate!(currentDay),
              dayNumber: currentDay.day,
              shortName:
                  shortName.length > 3 ? shortName.substring(0, 3) : shortName,
              onTap: () => _goToActualDay(index),
              isDimmed: currentDay
                  .isBefore(DateTime.now().subtract(Duration(days: 1))),
              available: widget.selectableDayPredicate == null
                  ? true
                  : widget.selectableDayPredicate!(currentDay),
              hasEvents: widget.dayWithEventsPredicate == null
                  ? false
                  : widget.dayWithEventsPredicate!(currentDay),
              scale: widget.scale,
              dotsColor: _dotsColor,
              dayColor: _dayColor,
              activeDayColor: _activeDayColor,
              activeDayBackgroundColor: _activeBackgroundDayColor,
            ),
          );
        },
      ),
    );
  }

  /// Creates the row with all the months in the calendar. If [widget.showYears] is set to true
  /// it will only show the months allowed in the selected year. By default it will show all
  /// months in the calendar and the small version of [YearItem] for each year in between
  Widget _buildMonthList() {
    return Container(
      height: 40 * widget.scale,
      child: ScrollablePositionedList.separated(
        initialScrollIndex: _monthSelectedIndex ?? 0,
        initialAlignment: _scrollAlignment,
        itemScrollController: _controllerMonth,
        padding: EdgeInsets.only(
          left: widget.offset,
          right: widget.offset,
        ),
        scrollDirection: Axis.horizontal,
        itemCount: _months.length,
        separatorBuilder: (_, __) => Divider(indent: widget.offset),
        itemBuilder: (BuildContext context, int index) {
          final currentDate = _months[index];
          final monthName = DateFormat.MMMM(_locale).format(currentDate);

          return Container(
            child: Row(
              mainAxisSize: MainAxisSize.min,
              mainAxisAlignment: MainAxisAlignment.center,
              crossAxisAlignment: CrossAxisAlignment.center,
              children: <Widget>[
                if (widget.firstDate.year != currentDate.year &&
                    currentDate.month == 1 &&
                    !widget.showYears)
                  Container(
                    padding: EdgeInsets.only(right: widget.offset),
                    child: YearItem(
                      name: DateFormat.y(_locale).format(currentDate),
                      color: _monthColor,
                      scale: widget.scale,
                      onTap: () {},
                    ),
                  ),
                MonthItem(
                  isSelected: _monthSelectedIndex == index,
                  name: monthName,
                  onTap: () => _goToActualMonth(index),
                  color: _monthColor,
                  scale: widget.scale,
                ),
              ],
            ),
          );
        },
      ),
    );
  }

  /// Creates the row with all the years in the calendar. It will only show if
  /// [widget.showYears] is set to true. It is false by default
  Widget _buildYearList() {
    return Container(
      height: 40 * widget.scale,
      child: ScrollablePositionedList.separated(
        initialScrollIndex: _yearSelectedIndex!,
        initialAlignment: _scrollAlignment,
        itemScrollController: _controllerYear,
        padding: EdgeInsets.only(
          left: widget.offset,
          right: widget.offset,
        ),
        scrollDirection: Axis.horizontal,
        itemCount: _years.length,
        separatorBuilder: (_, __) => Divider(indent: widget.offset),
        itemBuilder: (BuildContext context, int index) {
          final currentDate = _years[index];
          final yearName = DateFormat.y(_locale).format(currentDate);

          return Container(
            child: Row(
              mainAxisSize: MainAxisSize.min,
              mainAxisAlignment: MainAxisAlignment.center,
              crossAxisAlignment: CrossAxisAlignment.center,
              children: <Widget>[
                YearItem(
                  isSelected: _yearSelectedIndex == index,
                  onTap: () => _goToActualYear(index),
                  color: _monthColor,
                  name: yearName,
                  small: false,
                ),
                // if (index == _years.length - 1)
                //   SizedBox(
                //     width: MediaQuery.of(context).size.width -
                //         widget.leftMargin -
                //         (yearName.length * 10),
                //   )
              ],
            ),
          );
        },
      ),
    );
  }

  /// It will populate the [_days] list with all the allowed days. Adding all days of the month
  /// when the [selectedDate.month] is not the first or the last in [widget.firstDate] or [widget.lastDate].
  /// In that case it will only show the allowed days from and up to the specified in [widget.firstDate]
  /// and [widget.lastDate]
  _generateDays(DateTime? selectedDate) {
    _days.clear();
    for (var i = 1; i <= 31; i++) {
      final day = DateTime(selectedDate!.year, selectedDate.month, i);
      if (day.difference(widget.firstDate).inDays < 0) continue;
      if (day.month != selectedDate.month || day.isAfter(widget.lastDate))
        break;
      _days.add(day);
    }
  }

  /// It will populate the [_months] list. If [widget.showYears] is true, it will add from January
  /// to December, unless the selected year is the [widget.firstDate.year] or the [widget.lastDate.year].
  /// In that case it will only from and up to the allowed months in [widget.firstDate] and [widget.lastDate].
  /// By default, when [widget.showYears] is false, it will add all months from [widget.firstDate] to
  /// [widget.lastDate] and all in between
  _generateMonths(DateTime? selectedDate) {
    _months.clear();
    if (widget.showYears) {
      int month = selectedDate!.year == widget.firstDate.year
          ? widget.firstDate.month
          : 1;
      DateTime date = DateTime(selectedDate.year, month);
      while (date.isBefore(DateTime(selectedDate.year + 1)) &&
          date.isBefore(widget.lastDate)) {
        _months.add(date);
        date = DateTime(date.year, date.month + 1);
      }
    } else {
      DateTime date = DateTime(widget.firstDate.year, widget.firstDate.month);
      while (date.isBefore(widget.lastDate)) {
        _months.add(date);
        date = DateTime(date.year, date.month + 1);
      }
    }
  }

  /// It will populate the [_years] list with the years between firstDate and lastDate
  _generateYears() {
    _years.clear();
    DateTime date = widget.firstDate;
    while (date.isBefore(widget.lastDate)) {
      _years.add(date);
      date = DateTime(date.year + 1);
    }
  }

  /// It will reset the calendar to the initial date
  _resetCalendar(DateTime date) {
    if (widget.showYears) {
      _generateMonths(date);
      _moveToMonthIndex(_monthSelectedIndex ?? 0);
    }
    _generateDays(date);
    _daySelectedIndex = date.month == _selectedDate!.month
        ? _days.indexOf(
            _days.firstWhere((dayDate) => dayDate.day == _selectedDate!.day))
        : null;
    _moveToDayIndex(_daySelectedIndex ?? 0);
  }

  _goToActualYear(int index) {
    _moveToYearIndex(index);
    _yearSelectedIndex = index;
    _monthSelectedIndex = null;
    _resetCalendar(_years[index]);
    setState(() {});
  }

  void _moveToYearIndex(int index) {
    // _controllerYear.scrollTo(
    //   index: index,
    //   alignment: _scrollAlignment,
    //   duration: Duration(milliseconds: 500),
    //   curve: Curves.easeIn,
    // );
  }

  _goToActualMonth(int index) {
    _moveToMonthIndex(index);
    _monthSelectedIndex = index;
    _resetCalendar(_months[index]);
    setState(() {});
  }

  void _moveToMonthIndex(int index) {
    // _controllerMonth.scrollTo(
    //   index: index,
    //   alignment: _scrollAlignment,
    //   duration: Duration(milliseconds: 500),
    //   curve: Curves.easeIn,
    // );
  }

  _goToActualDay(int index) {
    _moveToDayIndex(index);
    _daySelectedIndex = index;
    _selectedDate = _days[index];
    widget.onDateSelected(_selectedDate);
    setState(() {});
  }

  void _moveToDayIndex(int index) {
    // _controllerDay.scrollTo(
    //   index: index,
    //   alignment: _scrollAlignment,
    //   duration: Duration(milliseconds: 500),
    //   curve: Curves.easeIn,
    // );
  }

  selectedYear() {
    _yearSelectedIndex = _years.indexOf(_years
        .firstWhere((yearDate) => yearDate.year == widget.initialDate.year));
  }

  selectedMonth() {
    if (widget.showYears)
      _monthSelectedIndex = _months.indexOf(_months.firstWhere(
          (monthDate) => monthDate.month == widget.initialDate.month));
    else
      _monthSelectedIndex = _months.indexOf(_months.firstWhere((monthDate) =>
          monthDate.year == widget.initialDate.year &&
          monthDate.month == widget.initialDate.month));
  }

  selectedDay() {
    _daySelectedIndex = _days.indexOf(
        _days.firstWhere((dayDate) => dayDate.day == widget.initialDate.day));
  }

  /// Initializes the calendar. It will be executed every time a new date is selected
  _initCalendar() {
    _selectedDate = widget.initialDate;
    if (widget.showYears) {
      _generateYears();
      selectedYear();
    }
    _generateMonths(_selectedDate);
    selectedMonth();
    _generateDays(_selectedDate);
    selectedDay();
  }
}


example/lib/main.dart
import 'package:calendar_timeline/calendar_timeline.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:flutter_localizations/flutter_localizations.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    SystemChrome.setSystemUIOverlayStyle(SystemUiOverlayStyle.light);
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Flutter Demo',
      theme: ThemeData(
        accentColor: Colors.blue,
        primarySwatch: Colors.blue,
        scaffoldBackgroundColor: Colors.white,
        backgroundColor: Color(0xFF333A47),
      ),
      localizationsDelegates: [
        GlobalMaterialLocalizations.delegate,
        GlobalWidgetsLocalizations.delegate,
        GlobalCupertinoLocalizations.delegate,
      ],
      supportedLocales: [
        const Locale('es'),
        const Locale('en'),
      ],
      home: HomePage(),
    );
  }
}

class HomePage extends StatefulWidget {
  @override
  _HomePageState createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  DateTime _selectedDate;

  @override
  void initState() {
    super.initState();
    _resetSelectedDate();
  }

  void _resetSelectedDate() {
    _selectedDate = DateTime.now();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Theme.of(context).backgroundColor,
      appBar: AppBar(
        elevation: 0,
        centerTitle: true,
        brightness: Brightness.dark,
        backgroundColor: Theme.of(context).backgroundColor,
        title: Text(
          'SELECTED DATE: $_selectedDate',
          style: TextStyle(fontSize: 16),
        ),
      ),
      body: Column(
        children: [
          CalendarTimeline(
            scale: 0.8,
            offset: 20,
            locale: 'ru',
            initialDate: _selectedDate,
            dayWithEventsPredicate: (date) => date.day != 14,
            // selectableDayPredicate: (date) => date.day != 23,
            firstDate: DateTime.now().subtract(Duration(days: 10)),
            lastDate: DateTime.now().add(Duration(days: 365)),
            onDateSelected: (date) {
              setState(() {
                _selectedDate = date;
              });
            },
            // dotsColor: Colors.white,
            // monthColor: Colors.white70,
            // dayColor: Colors.teal[200],
            // activeDayColor: Colors.white,
            // dayNameColor: Color(0xFF333A47),
            // activeBackgroundDayColor: Colors.redAccent[100],
          ),
          Expanded(
            child: Stack(
              children: [
                Container(
                  width: double.infinity,
                  margin: EdgeInsets.only(top: 8),
                  padding: EdgeInsets.only(top: 16),
                  decoration: BoxDecoration(
                    color: Colors.white,
                    borderRadius: BorderRadius.only(
                      topLeft: Radius.circular(16),
                      topRight: Radius.circular(16),
                    ),
                  ),
                  child: Column(
                    children: [
                      SizedBox(height: 20),
                      CupertinoButton(
                        color: Theme.of(context).backgroundColor,
                        child: Text(
                          'RESET',
                          style: TextStyle(color: Colors.white),
                        ),
                        onPressed: () => setState(
                          () => _resetSelectedDate(),
                        ),
                      ),
                    ],
                  ),
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }
}
